nbgrader/exchange/abc/exchange.py

```diff
@@ -0,0 +1,134 @@
+import os
+import datetime
+import sys
+
+from textwrap import dedent
+
+from dateutil.tz import gettz
+from traitlets.config import LoggingConfigurable
+from traitlets import Unicode, Bool, Instance, Type, default, validate
+from jupyter_core.paths import jupyter_data_dir
+
+from nbgrader.utils import check_directory, ignore_patterns, self_owned
+from nbgrader.coursedir import CourseDirectory
+from nbgrader.auth import Authenticator
+
+
+class ExchangeError(Exception):
+    pass
+
+
+class Exchange(LoggingConfigurable):
```

Comment:
It seems like some of the traits and methods of this class make some strong assumptions about the files being located in a unix-like filesystem (e.g. the 'root') traitlet. Perhaps it would better to move these to the default exchange implementation rather than having them in the base class?

Reply:
Yes, that's a good point. An easy way to get around this would be to list the options for the default exchange classes in the `NbGrader` app. Specifically, in `apps/baseapp.py`, you can have:

```
        # include all the exchange actions
        for ex_name in exchange.default.__all__:
            ex = getattr(exchange.default, ex_name)
            if hasattr(ex, "class_traits") and ex.class_traits(config=True):
                classes.append(ex)
```

on line 153 (i.e. instead of importing the classes from `exchange` you import them from `exchange.default`).

This will of course though not help if you are using a different exchange than the default one. But in that case, perhaps we can assume that since it is a custom thing you would be familiar with the available options?